recipe/conda_build_config.yaml

```diff
@@ -41,23 +41,38 @@ rust_default_cc:
   - arm64-apple-darwin20.0.0-clang      # [linux64 or (osx and x86_64) or (osx and arm64)]
   - clang-cl                            # [linux64 or (osx and x86_64) or win64]
 MACOSX_SDK_VERSION:
-  - 11.0
+  - 0                                   # [linux64]
+  - 0                                   # [linux64 or aarch64]
+  - 0                                   # [linux64 or ppc64le]
+  - 10.9                                # [linux64 or (osx and x86_64)]
+  - 11.0                                # [linux64 or (osx and x86_64) or (osx and arm64)]
+  - 0                                   # [linux64 or (osx and x86_64) or win64]
```

Comment:
Why do we need this?

Reply:
We needlessly download macOS SDKs for non-`osx` targets. E.g., https://cloud.drone.io/conda-forge/rust-activation-feedstock/40/2/2 failed because of a slow download for this.
Also, we probably want to have 10.9 instead of 11.0 for `osx-64`, right?

Reply:
11.0 is the SDK version, 10.9 is the deployment target

Reply:
I know, but AFAIK our current default is to use SDK=target version, unless specified otherwise.

----
Or do you mean to ask why there is a `MACOSX_SDK_VERSION` key here at all? I just kept it included because it was there in the first place. IIUC, we could remove it anyway (using the default SDK=target version via conda-forge-ci-setup).

Reply:
> IIUC, we could remove it anyway

Nope we can't because we need it for cross compiling.

> I know, but AFAIK our current default is to use SDK=target version, unless specified otherwise.

This is not a normal package and even has a canadian cross toolchain.

Reply:
> Nope we can't because we need it for cross compiling.

Why? Where/when does the build use/record `MACOSX_SDK_VERSION`?

> canadian cross toolchain.

Had to look this one up. What a terrible (=cryptic) name.

Reply:
> Why? Where/when does the build use/record MACOSX_SDK_VERSION?

It's used for downloading the SDK. for eg: we have the situation where, `(linux-64) build == host != target (osx-arm64)` in that case, we need the 11.0 SDK, but we can't remove `MACOS_SDK_VERSION` nor `MACOSX_DEPLOYMENT_TARGET` because it's needed for the cross compiling at test phase.

Reply:
But `cross_target_platform` is the key to decide when we'd need which SDK or target version, right?
Hence I added `MACOSX_SDK_VERSION`/`MACOSX_DEPLOYMENT_TARGET` to the `zip_keys` for `cross_target_platform`.
Meaning
- for every build with `cross_target_platform == "osx-64"` we'd get `MACOSX_DEPLOYMENT_TARGET == "10.9"`, and 
- for every build with `cross_target_platform == "osx-arm64"` we'd get `MACOSX_DEPLOYMENT_TARGET == "11.0"`
- for every build with `cross_target_platform not in {"osx-64", "osx-arm64"`}` nothing is set or downloaded.

With the logic from `conda-forge-ci-setup` we'd -- even if we'd remove `MACOSX_SDK_VERSION` from cbc.yaml -- `MACOSX_SDK_VERSION == "10.9"` (`cross_target_platform == "osx-64"`) or `MACOSX_SDK_VERSION == "11.0"` (`cross_target_platform == "osx-arm64"`).

Reply:
Theoretically you are correct, but practically as you've already realized we can't have more keys zipped due to a conda-smithy bug. I'm not going to merge this PR which would break rerendering.

Practically, setting `MACOSX_DEPLOYMENT_TARGET=10.9` on `osx-arm64` is equivalent to setting `MACOSX_DEPLOYMENT_TARGET=11.0`. So we can have `MACOSX_DEPLOYMENT_TARGET=10.9` everywhere without a problem.

Reply:
> With the logic from conda-forge-ci-setup we'd -- even if we'd remove MACOSX_SDK_VERSION from cbc.yaml -- MACOSX_SDK_VERSION == "10.9" (cross_target_platform == "osx-64") or MACOSX_SDK_VERSION == "11.0" (cross_target_platform == "osx-arm64")

This is wrong. If we remove `MACOSX_SDK_VERSION` from `cbc.yaml` it is inferred from `target_platform` which is not equal to `cross_target_platform` in this repo.

Reply:
> Theoretically you are correct, but practically as you've already realized we can't have more keys zipped due to a conda-smithy bug.

Yes, I'll submit a simple patch to smithy for this.

> I'm not going to merge this PR which would break rerendering.

Of course, I wouldn't expect you to merge this non-functional PR.

> Practically, setting `MACOSX_DEPLOYMENT_TARGET=10.9` on `osx-arm64` is equivalent to setting `MACOSX_DEPLOYMENT_TARGET=11.0`.

I was not aware of such special casing. Where is that handled?

> So we can have `MACOSX_DEPLOYMENT_TARGET=10.9` everywhere without a problem.

If/when we fixed the smithy bug, we can omit it (and thus the download) from non-osx-cross-targets.

Reply:
> If we remove `MACOSX_SDK_VERSION` from `cbc.yaml` it is inferred from `target_platform` 

Where can I find the logic/code that handles that?

Reply:
> Where can I find the logic/code that handles that?

https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/master/recipe/conda_build_config.yaml#L92

> I was not aware of such special casing. Where is that handled?

It's handled by the SDK itself. For osx-arm64 oldest support target is 11.0, so if we had 10.9, it'll automatically bump it to 11.0.

Reply:
> https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/master/recipe/conda_build_config.yaml#L92

This is the target version that we do override in the recipe's cbc.yaml. The SDK version is only inferred later on in conda-forge-ci-setup using only the deployment target version. Hence my reasoning.

> It's handled by the SDK itself. For osx-arm64 oldest support target is 11.0, so if we had 10.9, it'll automatically bump it to 11.0.

Ah, okay. Thanks. I didn't expect that.

Reply:
> The SDK version is only inferred later on in conda-forge-ci-setup using only the deployment target version. Hence my reasoning.

Nope. SDK version is read from the config file (`.ci_support/*.yml`) at https://github.com/conda-forge/conda-forge-ci-setup-feedstock/blob/master/recipe/download_osx_sdk.sh#L9

Reply:
> Nope. SDK version is read from the config file (`.ci_support/*.yml`) at https://github.com/conda-forge/conda-forge-ci-setup-feedstock/blob/master/recipe/download_osx_sdk.sh#L9

But only if it's set!? Which it wouldn't if we don't add it to cbc.yaml

Reply:
True. There's also the need to have SDK 11 when you are doing `target_platform == osx-arm64` and `cross_target_platform == osx-64` where if we go by `cross_target_platform` we'll only use SDK 10.9, but in order to cross compile some of the rust packages in test, you need to build native binaries which require SDK 11.

It's a lot to digest, so don't hesitate to answer any questions you may have. I'm happy to answer.

Reply:
On second thought, we don't have that requirement yet, because we are cross-compiling arm64. We might have the need in the future if we have arm64 native CI

Reply:
> It's a lot to digest, so don't hesitate to answer any questions you may have. I'm happy to answer.

No, everything makes sense thus far and you've already invested enough of your time today to go through this with me ;). Many thanks for that!